linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/Emitter.scala

```diff
@@ -544,51 +578,58 @@ final class Emitter(config: Emitter.Config, prePrinter: Emitter.PrePrinter) {
         val ctorCache = classCache.getConstructorCache()
 
         if (isJSClass) {
-          assert(linkedInlineableInit.isEmpty)
+          assert(linkedInlineableInit_!.isEmpty) // just an assertion, it's fine
 
-          val jsConstructorDef = linkedClass.jsConstructorDef.getOrElse {
+          // Explicitly versioned just below
+          val jsConstructorDef = linkedClass_!.jsConstructorDef.getOrElse {
             throw new IllegalArgumentException(s"$className does not have an exported constructor")
           }
 
-          val ctorVersion = Version.combine(linkedClass.version, jsConstructorDef.version)
+          val ctorVersion = Version.combine(classVersion, jsConstructorDef.version)
           ctorCache.getOrElseUpdate(ctorVersion,
               classEmitter.genJSConstructor(
-                className, // invalidated by overall class cache (part of ancestors)
-                linkedClass.superClass, // invalidated by class version
-                hasJSSuperClass, // invalidated by class version
-                useESClass, // invalidated by class version
+                className, // always safe
+                linkedClass_!.superClass, // invalidated by class version
+                linkedClass_!.jsSuperClass.isDefined, // invalidated by class version (*not* local decision-making)
+                useESClass, // always safe
                 jsConstructorDef // part of ctor version
-              )(moduleContext, ctorCache, linkedClass.pos).map(prePrint(_, memberIndent)))
+              )(moduleContext, ctorCache, linkedClass_!.pos) // pos invalidated by class version
+                .map(prePrint(_, memberIndent))
+          )
         } else {
-          val ctorVersion = linkedInlineableInit.fold {
-            Version.combine(linkedClass.version)
+          val ctorVersion = linkedInlineableInit_!.fold { // versioning
```

Comment:
I wonder if we should move calculation of this version to `extractInlineableInit` (and return a 3-tuple).

Reply:
I don't think so. It would obscure the reason why the other "invalidated by class version" are correct (`superClass, `pos`).